### PR TITLE
fix wandb property name

### LINKED ISSUE #12603: WandbLogger name return the project name rather than the run/exp name
## 🐛 Bug

`self._experiment.project_name()` is not the `self._name` but `self._project`

https://github.com/PyTorchLightning/pytorch-lightning/blob/7d1075d2116f8ceda9f7bf12cc083fbc1eb0f32e/pytorch_lightning/loggers/wandb.py#L450

### To Reproduce

```ipython
In [1]: import wandb
   ...:
   ...: run = wandb.init(project="lightning-template", entity="wangtianshu", name="run-name")
Failed to detect the name of this notebook, you can set it manually with the WANDB_NOTEBOOK_NAME environment variable to enable code saving.
wandb: Currently logged in as: wangtianshu (use `wandb login --relogin` to force relogin)
wandb: Tracking run with wandb version 0.12.11
wandb: Run data is saved locally in /home1/wangtianshu/lightning-template/wandb/run-20220404_171224-1z8az7jn
wandb: Run `wandb offline` to turn off syncing.
wandb: Syncing run run-name
wandb: ⭐️ View project at https://wandb.ai/wangtianshu/lightning-template
wandb: 🚀 View run at https://wandb.ai/wangtianshu/lightning-template/runs/1z8az7jn

In [2]: run.project_name()
Out[2]: 'lightning-template'

In [3]: run.name
Out[3]: 'run-name'
```

### Expected behavior

`wandb_logger.name` return then correct name rather than project name.

### Environment

* Packages:
	- numpy:             1.22.2
	- pyTorch_debug:     False
	- pyTorch_version:   1.11.0+cu113
	- pytorch-lightning: 1.6.0
	- tqdm:              4.63.0
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- ELF
	- processor:         x86_64
	- python:            3.9.7
	- version:           #177-Ubuntu SMP Thu Feb 3 10:50:38 UTC 2022

cc @awaelchli @morganmcg1 @AyushExel @borisdayma @scottire @manangoel99

### PR DESCRIPTION
**Pull request by tshu-w**
## What does this PR do?

Fixes #12603 This fix the name property of WandbLogger for https://github.com/PyTorchLightning/pytorch-lightning/issues/12028.

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by manangoel99** [comment]
Hello @tshu-w ! This fix would be awesome!

**Comment by tshu-w** [comment]
@manangoel99 Thank you for confirming and link the old PR. Let's ping PL members to make sure this PR is perfected and merged if there is no reply for some time.

**Comment by krshrimali** [comment]
Hi, @tshu-w - Thank you for creating this PR. We appreciate the contribution 🚀 Looks like the tests are failing, mostly because there is `project_name()` used in the tests which should be renamed as well to `name`. Would you mind updating the relevant tests and ping once ready? Please see the log here: https://github.com/PyTorchLightning/pytorch-lightning/runs/5813691929?check_suite_focus=true.

Apologies for the delay in the old PR!

**Comment by manangoel99** [comment]
This will bring one change though. The directory structure for checkpoints currently looks like `<project_name>/<run_id>` but after this it will be `<run_name>/<run_id>`

**Comment by manangoel99** [comment]
@tshu-w though having `logger.name` return the run name makes more sense but the changed directory structure does not seem as intuitive. So maybe instead of chaning the behaviour of the `name` method we could have a new method `run_name` which returns the run name. What do you think?

**Comment by tshu-w** [comment]
@krshrimali Hi, thank you remind, I update the test file now.

> This will bring one change though. The directory structure for checkpoints currently looks like `<project_name>/<run_id>` but after this it will be `<run_name>/<run_id>`

This is as expected, PL default `TensorBoardLogger` stores checkpoints to `save_dir/name/version`(for more discuss, see https://github.com/PyTorchLightning/pytorch-lightning/issues/12028), and I believe that dir, name and id in wandb correspond to save_dir, name and version respectively.

> @tshu-w though having `logger.name` return the run name makes more sense but the changed directory structure does not seem as intuitive. So maybe instead of chaning the behaviour of the `name` method we could have a new method `run_name` which returns the run name. What do you think?

As mentioned above, the save_dir, name, and version of a PL logger have specific meanings, and I think the wandb name is the concept that just corresponds to the PL logger's name. Therefore, we need to change the behavior of name, rather than defining a new run_name (which could also be an alias for name, but that's beyond this PR).
cc @awaelchli

**Comment by manangoel99** [comment]
QQ what does name in PL signify. What I mean by this is that you could be naming a project or an experiment in the project. 
The `name` property in WandbLogger assumes that the user is naming the project and not the experiment (equivalent to a wandb run object).

**Comment by tshu-w** [comment]
> QQ what does name in PL signify. What I mean by this is that you could be naming a project or an experiment in the project.
> The `name` property in WandbLogger assumes that the user is naming the project and not the experiment (equivalent to a wandb run object).

I'm the new users of wandb, but from the document here https://docs.wandb.ai/ref/python/init, should `project` for project name and `name` for run/experiment name? And I see using `run.name` got the `name` in init_args while `project_name()` got the `project`. Is there something I'm getting wrong?

As for which `name` in PL logger should correspond, I believe it is run name.

**Comment by manangoel99** [comment]
@tshu-w we can make this change but if `save_dir=None`, this causes a lot of clutter for the user so we can change it to have `save_dir=self.experiment.project_name()` if `save_dir=None` so that everything goes under one directory.
We would also like to keep a deprecation warning here to make sure it doesn't break any existing production workflows.
cc @awaelchli

**Comment by tshu-w** [comment]
@manangoel99 agree with you, but that's beyond the scope of this PR.

**Comment by manangoel99** [comment]
@tshu-w I think in that case maybe we could close this PR and I can come up with another PR meanwhile to have deprecation warnings now since the behaviour of the arguments will be changing and this could break people's code so we need to raise deprecation warnings.
@awaelchli apologies if this is an ignorant question but does pytorch lightning have guidelines for handling changes in argument behaviour. Something like keeping it the same in 1.7 with warnings and then finally changing it in 1.8.

**Comment by tshu-w** [comment]
> I think in that case maybe we could close this PR and I can come up with another PR meanwhile to have deprecation warnings now since the behaviour of the arguments will be changing and this could break people's code so we need to raise deprecation warnings.

It seems to me that the inconsistency between `self._experiment.project_name()` and `self._name` can be solved first (as has been suggested by many people). If the user is bothered by multiple directories, this can also be solved by simply setting `save_dir`. Anyway, let's hear from PL members.

**Comment by awaelchli** [comment]
May I suggest the following. 

1. Return the run name in the name property as done in this PR.
2. Since the project name is mandatory to provide, and the name optional, set the run name to the project name by default (instead of letting wandb generate random names). 

This way, we are backward compatible partially, and the directory doesn't get cluttered every time you run a new experiment.

**Comment by tshu-w** [comment]
@manangoel99 @awaelchli suggest make sense to me, what about you?

**Comment by manangoel99** [comment]
@awaelchli thanks for the suggestion and @tshu-w  I think we can go ahead with this for now but in the long term I think it would be better if we set the `save_dir` to the project by default. 
So we can go ahead with @awaelchli's suggestion and also add a deprecation warning that the new behaviour will change in the next release to what I just mentioned above.

**Comment by awaelchli** [comment]
Hey @tshu-w would you like to continue this PR as we discussed?

**Comment by tshu-w** [comment]
> Hey @tshu-w would you like to continue this PR as we discussed?

Hi，I’m really busy this two weeks, maybe next week.

**Comment by tshu-w** [comment]
> 2\. Since the project name is mandatory to provide, and the name optional, set the run name to the project name by default (instead of letting wandb generate random names).

@awaelchli Hi, this https://github.com/PyTorchLightning/pytorch-lightning/pull/12037 wants to set default save_dir and name, should I do this or as your suggest.

**Comment by awaelchli** [comment]
@tshu-w, yes I believe we should continue this PR here with the idea we discussed.

**Comment by tshu-w** [comment]
@awaelchli Hi, I have updated it. Can you take a look.

**Review by awaelchli — COMMENTED** [review]
Looks great. Can we keep one of the tests that asserts the project name on the experiment when name is not provided? Thanks!

File: CHANGELOG.md [hunk 1]
```diff
@@ -214,6 +214,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Don't raise a warning when `nn.Module` is not saved under hparams ([#12669](https://github.com/PyTorchLightning/pytorch-lightning/pull/12669))
 - Raise `MisconfigurationException` when the accelerator is available but the user passes invalid `([]/0/"0")` values to the `devices` flag ([#12708](https://github.com/PyTorchLightning/pytorch-lightning/pull/12708))
 - Support `auto_select_gpus` with the accelerator and devices API ([#12608](https://github.com/PyTorchLightning/pytorch-lightning/pull/12608))
+- Fix the `name` property of `WandbLogger` ([#12157](https://github.com/PyTorchLightning/pytorch-lightning/issues/12157))
```
[the remark this answers is not in the document]
Comment (tshu-w) [reply CHANGELOG.md]: Done, please check it.

**Comment by tshu-w** [comment]
> Looks great. Can we keep one of the tests that asserts the project name on the experiment when name is not provided? Thanks!

Hi, I found it hard to keep one of the tests because my modifications are at the time of initialization, not the `return` in the test. I made some changes to the test, which you can check at my last commit https://github.com/PyTorchLightning/pytorch-lightning/pull/12604/commits/4e500bac3c2c94e29bf7a72bc44fbc94c68670e0.

File: CHANGELOG.md [hunk 2]
```diff
@@ -91,6 +91,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Raise an error if there are insufficient training batches when using a float value of `limit_train_batches` ([#12885](https://github.com/PyTorchLightning/pytorch-lightning/pull/12885))
 
 
+- The `WandbLogger` will now use the run name in the logs folder if it is provided, and otherwise the project name  ([#12157](https://github.com/PyTorchLightning/pytorch-lightning/issues/12157))
```
Comment (awaelchli) [inline CHANGELOG.md:94]:
```suggestion
- The `WandbLogger` will now use the run name in the logs folder if it is provided, and otherwise the project name ([#12157](https://github.com/PyTorchLightning/pytorch-lightning/issues/12157))
```

**Comment by awaelchli** [comment]
@tshu-w this is simple to do. I can't help you directly unfortunately, because you opened this pr directly from master (normally you should create a separate branch in your fork).
Please add these lines in tests/loggers/test_wandb.py LINE 45

```py
    wandb.init.reset_mock()

    # test wandb.init called with project as name if name not provided
    WandbLogger(project="test_project")
    wandb.init.assert_called_once_with(
        name="test_project", dir=None, id=None, project="test_project", resume="allow", anonymous=None
    )
```

after that LGTM

**Comment by tshu-w** [comment]
@awaelchli Thanks for the detailed instructions, I got the tests added. Next time I will use the new branch to PR.

**Comment by tshu-w** [comment]
@awaelchli Would you please look at it, otherwise there will be a new conflict soon. ;)

**Review by awaelchli — APPROVED** [review]

**Comment by rohanshad** [comment]
I just noticed my wandb logger folder setup is all messed up with PL 1.6.3, looking forward to pulling this fix when it's merged!

**Comment by tshu-w** [comment]
> I just noticed my wandb logger folder setup is all messed up with PL 1.6.3, looking forward to pulling this fix when it's merged!

Hi, this [issue](https://github.com/PyTorchLightning/pytorch-lightning/issues/12157) exists in PL 1.5 either, If you meet error in PL 1.6.3. It might be other bug?

**Comment by rohanshad** [comment]
> > I just noticed my wandb logger folder setup is all messed up with PL 1.6.3, looking forward to pulling this fix when it's merged!
> 
> Hi, this [issue](https://github.com/PyTorchLightning/pytorch-lightning/issues/12157) exists in PL 1.5 either, If you meet error in PL 1.6.3. It might be other bug?
> 
> And if @awaelchli could at other appropriate reviewers.

On 1.5.10 the logger would store checkpoints like this:
```
project_name
        ├── version_id (alphanumeric 8 digit)
        ├── version_id (alphanumeric 8 digit)
                          ├── checkpoints
```

On 1.6.3 they're all like this now:

```
version_name **(not the same as version_id)**
        ├── version_None 
                        ├── checkpoints
```

Completely messes up the way I track experiments since on the online interface of wandb references **version_id** and not version_name. Is this a different issue altogether?

**Comment by tshu-w** [comment]
> Completely messes up the way I track experiments since on the online interface of wandb references **version_id** and not version_name. Is this a different issue altogether?

I suspect yes, though I'm not why this happening on your side. This PR just want to change logger structure like this:

```
run_name or project_name
        ├── version_id (alphanumeric 8 digit)
        ├── version_id (alphanumeric 8 digit)
                          ├── checkpoints
```

**Comment by awaelchli** [comment]
Yes I pinged the team for one the third review. Sorry about the wait time, things are slow at the moment.

File: CHANGELOG.md [hunk 3]
```diff
@@ -97,6 +97,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Changed `pytorch_lightning.core.lightning` to `pytorch_lightning.core.module` ([#12740](https://github.com/PyTorchLightning/pytorch-lightning/pull/12740))
 
 
+- The `WandbLogger` will now use the run name in the logs folder if it is provided, and otherwise the project name  ([#12157](https://github.com/PyTorchLightning/pytorch-lightning/issues/12157))
```
Comment (akihironitta) [inline CHANGELOG.md:100]:
```suggestion
- The `WandbLogger` will now use the run name in the logs folder if it is provided, and otherwise the project name  ([#12604](https://github.com/PyTorchLightning/pytorch-lightning/pull/12604))
```
